jgrapht-core/src/main/java/org/jgrapht/alg/MinSourceSinkCut.java

```diff
@@ -30,116 +30,143 @@
-
+        //Compute a maxFlow from source to sink
+        MaximumFlow<E> maxFlow = maxFlowAlg.buildMaximumFlow(source, sink);
         this.cutWeight = maxFlow.getValue();
 
+        if(graph instanceof DirectedGraph)
+            processDirectedGraph(maxFlow);
+        else
+            processUndirectedGraph(maxFlow);
+    }
+
+    private void processDirectedGraph( MaximumFlow<E> maxFlow){
+        minCut=new LinkedHashSet<>();
+        DirectedGraph<V,E> directedGraph=(DirectedGraph<V,E>)graph;
         Queue<V> processQueue = new LinkedList<>();
         processQueue.add(source);
 
         while (!processQueue.isEmpty()) {
             V vertex = processQueue.remove();
-            if (minCut.contains(vertex)) {
+            if (minCut.contains(vertex))
                 continue;
-            } else {
-                minCut.add(vertex);
-            }
+
+            minCut.add(vertex);
 
             //1. Get the forward edges with residual capacity
-            Set<E> outEdges = new HashSet<>(graph.outgoingEdgesOf(vertex));
-            for (Iterator<E> it = outEdges.iterator(); it.hasNext();) {
-                E edge = it.next();
-                double edgeCapacity = graph.getEdgeWeight(edge);
+            for (E edge : directedGraph.outgoingEdgesOf(vertex)) {
+                double edgeCapacity = directedGraph.getEdgeWeight(edge);
                 double flowValue = maxFlow.getFlow().get(edge);
-                if (Math.abs(edgeCapacity - flowValue) <= epsilon) { //No residual capacity on the edge
-                    it.remove();
+                if (edgeCapacity - flowValue >= epsilon) { //Has some residual capacity left
+                    processQueue.add(directedGraph.getEdgeTarget(edge));
                 }
             }
-            for (E edge : outEdges) {
-                processQueue.add(Graphs.getOppositeVertex(graph, edge, vertex));
-            }
 
             //2. Get the backward edges with non-zero flow
-            Set<E> inEdges = new HashSet<>(graph.incomingEdgesOf(vertex));
-            for (Iterator<E> it = inEdges.iterator(); it.hasNext();) {
-                E edge = it.next();
-
-                //double edgeCapacity=graph.getEdgeWeight(edge);
+            for (E edge : directedGraph.incomingEdgesOf(vertex)) {
                 double flowValue = maxFlow.getFlow().get(edge);
-                if (flowValue <= epsilon) { //There is no flow on this edge
-                    it.remove();
+                if (flowValue >= epsilon) { //Has non-zero flow
+                    processQueue.add(directedGraph.getEdgeSource(edge));
                 }
             }
-            for (E edge : inEdges) {
-                processQueue.add(Graphs.getOppositeVertex(graph, edge, vertex));
+        }
+    }
+
+    public void processUndirectedGraph(MaximumFlow<E> maxFlow){
+//        TODO check implementation flow algorithm for undirected graphs. Seems wrong
```

Comment:
Hm, this is the wrong version. Must have messed something up with version control. Will revert this file back to the original.